Component Browser Section Navigator (left bar)

vitvakatu opened the pull request:
### Pull Request Description

[ci no changelog needed]
[Task link](https://www.pivotaltracker.com/story/show/181433641)

This PR adds a left bar with section navigation buttons to the Searcher List Panel. The buttons are implemented as a list view with an adjusted style.

https://user-images.githubusercontent.com/6566674/179517378-dba00f41-d32e-4ffb-a4d9-3cd376b3e781.mp4


### Important Notes

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

vitvakatu committed:
Add icons, background shape and place them on scene

vitvakatu committed:
Draft implementation finished

vitvakatu committed:
Adjust icons

vitvakatu committed:
Refactoring shapes & adjusting buttons positions

vitvakatu committed:
Adjusting icon positions

vitvakatu committed:
Adjusting background positions

vitvakatu committed:
Refactoring & icon colors

vitvakatu committed:
Refactoring & self-review

vitvakatu committed:
Apply enso-formatter

vitvakatu committed:
Clippy fixes

vitvakatu committed:
Add comments about rendering issues & bump wasm-size-limit

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/left-bar-181433641

vitvakatu committed:
Fix comment formatting

wdanilo commented on app/gui/view/component-browser/component-group/src/icon.rs:
```diff
@@ -98,10 +99,48 @@ crate::define_icons! {
+            pointer_events = false;
+            (style: Style, strong_color: Vector4, weak_color: Vector4) {
+                let dot = Circle(2.0.px());
+                let outer = Circle(6.5.px()) - Circle(5.5.px());
```
this is design error. Make these full values of 7 and 6.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -380,10 +419,15 @@ pub struct Model {
+    // FIXME[#182593513]: This is a hack to mitigate shadow rendering issues. Remove when the
+    //   issue is fixed and replace with a shadow embedded into the [`background`] shape.
+    //   See https://www.pivotaltracker.com/story/show/182593513.
+    navigator_shadow:    navigator_shadow::View,
```
This doesnt explain "what is the hack". I don't feel the next line looks like a hack. Please explain that better.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -488,6 +547,21 @@ impl Model {
     }
+
+    /// Scroll to the bottom of the [`section`].
+    fn show_section(&self, section: Section, style: &Style) {
```
How "show_section" "scrolls to the bottom of the section"? I feel that the function name and its docs are two different things.

vitvakatu replied:
Renamed to `scroll_to`

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/navigator.rs:
```diff
@@ -0,0 +1,255 @@
+//! The section navigator bar. This is a narrow bar on the left of the Searcher List Panel that
+//! contains two sets of navigation buttons.
+//!
+//! See [`Navigator`] struct.
```
I really ike doc comments with nice and clear explanations in this PR. Here, please link the design doc too.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/navigator.rs:
```diff
@@ -0,0 +1,255 @@
+pub enum Section {
+    SubModules,
+    LocalScope,
+    Favourites,
```
Please use American English spelling

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/navigator.rs:
```diff
@@ -0,0 +1,255 @@
+    fn default() -> Self {
+        Section::Favourites
+    }
+}
```
derive it

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/navigator.rs:
```diff
@@ -0,0 +1,255 @@
+    pub chosen_section: frp::Source<Option<Section>>,
+}
+
+const TOP_BUTTONS: [icon::Id; 2] = [icon::Id::Libraries, icon::Id::Marketplace];
+const BOTTOM_BUTTONS: [icon::Id; 3] = [icon::Id::SubModules, icon::Id::LocalScope, icon::Id::Star];
```
We've been chatting that these buttons should be placed in an auto-layout. I dont see an auto-layout implementation as part of this PR neither explanation that it will be added later. CC @farmaazon

farmaazon replied:
> We've been chatting that these buttons should be placed in an auto-layout. I dont see an auto-layout implementation as part of this PR neither explanation that it will be added later. CC @farmaazon

We decided to skip layouters for now, because our timeline is very tight.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/navigator.rs:
```diff
@@ -0,0 +1,255 @@
+        bottom_buttons.show_background_shadow(false);
+        display_object.add_child(&top_buttons);
+        display_object.add_child(&bottom_buttons);
+        // Top buttons are disabled at the moment.
```
"at the moment" does not explain why and when they will not be disabled.

vitvakatu replied:
Added a link to the pivotal task

farmaazon commented on app/gui/view/component-browser/searcher-list-panel/src/navigator.rs:
```diff
@@ -0,0 +1,255 @@
+
+
+
+// ============
+// === Icon ===
+// ============
```
The Icon model for list view deserves being more public - perhaps in a submodule of icon module.

vitvakatu replied:
Refactored

vitvakatu committed:
Address review comments

vitvakatu committed:
formatting

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/left-bar-181433641

akavel reviewed (changes requested):
Discrepancies from Design Doc & Task Description noticed during Acc/QA Review:

**1. No caption is shown when hovering the mouse over the marketplace icons.**

The Design Doc [requires this functionality](https://github.com/enso-org/design/blob/main/epics/component-browser/design.md#the-marketplace-mode) and it's explicitly called out in the task's "Notes" section.

**2. Marketplace icons are blue, should be gray.**

See the images in the Design Doc, incl. [the one below](https://github.com/enso-org/design/blob/main/epics/component-browser/design.md#the-marketplace-mode), as well as [the updated design posted in the Discord thread](https://discord.com/channels/401396655599124480/986214688163495956).

![](https://user-images.githubusercontent.com/1623053/150731604-17a1f3ca-fe1f-44a3-abe9-35ef010c7796.png)

**3. Bad UX: The selection jumps between items when the mouse is hovered nearby.**

The correct UX here is arguable; I'm reporting it as personally I found it unexpected and distracting while testing the feature. See the screencast below for a visual explanation (sorry if the textual description is not clear, I'm not sure how to better describe it):

https://user-images.githubusercontent.com/273837/177299026-50ea2ed7-21d7-46d0-a11b-2af302b950cc.mov

**4. Bad UX: Selection does not switch between items when scrolling the Component List Panel.**

When scrolling the Component List Panel to a different section, I would expect the selection highlight in the Section Navigator to jump to the new section, helping me notice that I switched the sections, and providing visual feedback teaching me the meaning of sections in the Section Navigator.

**5. Order of icons does not follow the updated design from 2022-06-14.**

See the [thread on Discord](https://discord.com/channels/401396655599124480/986214688163495956): bottom two icons should have their order switched.

**6. Cannot navigate by keyboard.**

The Task Description mentions: "*I can quickly change section with keyboard or mouse.*" The Design Doc also mentions keyboard navigation for the Section Navigator. I was not able to get it to work while testing.

farmaazon commented:
> Discrepancies from Design Doc & Task Description noticed during Acc/QA Review:
> 
> **1. No caption is shown when hovering the mouse over the marketplace icons.**

This, as I know, will be implemented in a separate issue (https://www.pivotaltracker.com/story/show/182613789), so probably this task description should be amended.

> **6. Cannot navigate by keyboard.**
> 

As I know, the keyboard navigation is implemented (because the section navigators are a list view, and list views have the navigation). It just has to be focused, but managing focus is in scope of another user story. So the task description should be updated, to not require keyboard navigation right now.


<blockquote><div><strong><a href="https://www.pivotaltracker.com/story/show/182613789">A caption when hovering the marketplace button on the left bar of CB - Pivotal Tracker</a></strong></div></blockquote>

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/left-bar-181433641

# Conflicts:
#	app/gui/view/component-browser/searcher-list-panel/src/lib.rs
#	build-config.yaml

vitvakatu committed:
Fixes after merge & apply enso-formatter

vitvakatu committed:
Fix colors of libraries and marketplace icons according to design doc

vitvakatu committed:
Disable selection movement when hovering entries with mouse

vitvakatu committed:
Revert "Disable selection movement when hovering entries with mouse"

This reverts commit 8ab58ffeab5c044f4fbbae702175ed5498873bca.

vitvakatu committed:
Disable selection movement when hovering entries with mouse, implement selecting section navigator buttons when scrolling

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/left-bar-181433641

# Conflicts:
#	Cargo.lock
#	app/gui/view/component-browser/searcher-list-panel/src/column_grid.rs
#	app/gui/view/component-browser/searcher-list-panel/src/lib.rs

vitvakatu committed:
Fix width of the wide component group section

vitvakatu committed:
Fix zooming behavior of the component browser demo scene

vitvakatu committed:
Minor changes in formatting

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/left-bar-181433641

vitvakatu commented:
@akavel the PR was updated.
1.  No caption is shown when hovering the mouse over the marketplace icons.
As @farmaazon mentioned, this is a part of a separate task

2. Marketplace icons are blue, should be gray.

Thanks, fixed (see updated screencast)

3. Bad UX: The selection jumps between items when the mouse is hovered nearby.

Fixed, now the buttons do not react on hovering with the mouse at all, you can only click on them.

4. Bad UX: Selection does not switch between items when scrolling the Component List Panel.

Fixed, the selection now follows the currently visible section (to be more precise, the bottom-most visible section). There is a minor visual gitch there though - the section jumps awkwardly when moving from Favorites to Submodules, and the fix is not trivial.

5. Order of icons does not follow the updated design from 2022-06-14.

The order of icons matches the section's order. We can change it easily when we change the order of sections.

6. Cannot navigate by keyboard.

As @farmaazon correctly noted, this is implemented, but we can't use it before implementing Focus controls.


Updated screencast: 


https://user-images.githubusercontent.com/6566674/179482275-df6ea675-7293-4a84-81b4-d2eed994fd14.mp4

vitvakatu committed:
Fix bug with selection jumps

Expose additional API in scroll area

vitvakatu commented:
Fixed the bug with selection jumps, see updated screencast in the PR description

vitvakatu committed:
Fix tracing API usage in nodes.rs

It was not working correctly

farmaazon commented on lib/rust/ensogl/component/scroll-area/src/lib.rs:
```diff
@@ -278,6 +282,8 @@ impl ScrollArea {
 
             frp.source.scroll_position_x <+ model.h_scrollbar.thumb_position.map(|x| -x);
             frp.source.scroll_position_y <+ model.v_scrollbar.thumb_position;
+            frp.source.scroll_position_target_x <+ model.h_scrollbar.thumb_position_target.map(|x| -x);
+            frp.source.scroll_position_target_y <+ model.v_scrollbar.thumb_position_target;
```
Are you sure you've covered all cases? What if scroll position was set by `scroll_to_x` or `scroll_to_y`?

vitvakatu replied:
I'm pretty sure. `scroll_to_x` leads to `scrollbar.scroll_to`, which in turn sets the `thumb_position_target` here:
```rust
unbounded_target_position <- any(&frp.scroll_to,&frp.jump_to);
thumb_position.target     <+ all_with3(&unbounded_target_position,&frp.set_thumb_size,
                &frp.set_max,|target,&size,&max| target.min(max-size).max(0.0));
thumb_position.skip       <+ frp.jump_to.constant(());
frp.source.thumb_position_target <+ thumb_position.target;
frp.source.thumb_position <+ thumb_position.value;
```

farmaazon commented on app/gui/view/debug_scene/component-list-panel-view/src/lib.rs:
```diff
@@ -171,6 +171,7 @@ fn init_local_cope_section(searcher_list_panel: &ComponentBrowserPanel) {
 #[entry_point]
 #[allow(dead_code)]
 pub fn main() {
+    init_tracing(DEBUG);
```
Please roll back to WARN.

vitvakatu replied:
Done

farmaazon commented on app/gui/view/component-browser/searcher-list-panel/src/navigator.rs:
```diff
@@ -2,6 +2,8 @@
 //! contains two sets of navigation buttons.
 //!
 //! See [`Navigator`] struct.
+//!
+//! See the [Component Browser Design Document](https://github.com/enso-org/design/blob/e6cffec2dd6d16688164f04a4ef0d9dff998c3e7/epics/component-browser/design.md).
```
Two paragraphs starting with `See...` look strange. Change some, or merge it to one sentence.

vitvakatu replied:
Done

vitvakatu committed:
Minor review fixes

vitvakatu committed:
Use additional layer to ensure proper shapes ordering

vitvakatu committed:
Bump wasm size limit

vitvakatu committed:
Fix layouting issue and an issue with invalid section selected when reopening the component browser

akavel reviewed (approved):
Acceptance/QA Review LGTM.

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/left-bar-181433641

# Conflicts:
#	build-config.yaml
#	lib/rust/frp/src/nodes.rs

vitvakatu committed:
Bump wasm size limit after merge with develop

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/left-bar-181433641
